// @ngx-formly/bootstrap/esm5/lib/bootstrap.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { FormlyModule } from '@ngx-formly/core';
import { FormlySelectModule } from '@ngx-formly/core/select';
import { BOOTSTRAP_FORMLY_CONFIG, FIELD_TYPE_COMPONENTS } from './bootstrap.config';
import { FormlyBootstrapAddonsModule } from '@ngx-formly/bootstrap/addons';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './types/input';
import * as ɵngcc2 from './types/checkbox';
import * as ɵngcc3 from './types/radio';
import * as ɵngcc4 from './types/select';
import * as ɵngcc5 from './types/textarea';
import * as ɵngcc6 from './types/multicheckbox';
import * as ɵngcc7 from './wrappers/form-field.wrapper';
import * as ɵngcc8 from '@ngx-formly/core';
var FormlyBootstrapModule = /** @class */ (function () {
    function FormlyBootstrapModule() {
    }
FormlyBootstrapModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: FormlyBootstrapModule });
FormlyBootstrapModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function FormlyBootstrapModule_Factory(t) { return new (t || FormlyBootstrapModule)(); }, imports: [[
            CommonModule,
            ReactiveFormsModule,
            FormlySelectModule,
            FormlyModule.forChild(BOOTSTRAP_FORMLY_CONFIG),
            FormlyBootstrapAddonsModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FormlyBootstrapModule, { declarations: function () { return [ɵngcc1.FormlyFieldInput, ɵngcc2.FormlyFieldCheckbox, ɵngcc3.FormlyFieldRadio, ɵngcc4.FormlyFieldSelect, ɵngcc5.FormlyFieldTextArea, ɵngcc6.FormlyFieldMultiCheckbox, ɵngcc7.FormlyWrapperFormField]; }, imports: function () { return [CommonModule,
        ReactiveFormsModule,
        FormlySelectModule, ɵngcc8.FormlyModule, FormlyBootstrapAddonsModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormlyBootstrapModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    FIELD_TYPE_COMPONENTS,
                ],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    FormlySelectModule,
                    FormlyModule.forChild(BOOTSTRAP_FORMLY_CONFIG),
                    FormlyBootstrapAddonsModule,
                ]
            }]
    }], function () { return []; }, null); })();
    return FormlyBootstrapModule;
}());
export { FormlyBootstrapModule };
