// @ngx-formly/material/bundles/ngx-formly-material.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@ngx-formly/material/form-field'),require('@ngx-formly/material/input'),require('@ngx-formly/material/textarea'),require('@ngx-formly/material/radio'),require('@ngx-formly/material/checkbox'),require('@ngx-formly/material/multicheckbox'),require('@ngx-formly/material/select'),exports, require('@angular/core'), require('@ngx-formly/material/form-field'), require('@ngx-formly/material/input'), require('@ngx-formly/material/textarea'), require('@ngx-formly/material/radio'), require('@ngx-formly/material/checkbox'), require('@ngx-formly/material/multicheckbox'), require('@ngx-formly/material/select')) :
    typeof define === 'function' && define.amd ? define('@ngx-formly/material', ['@angular/core','@ngx-formly/material/form-field','@ngx-formly/material/input','@ngx-formly/material/textarea','@ngx-formly/material/radio','@ngx-formly/material/checkbox','@ngx-formly/material/multicheckbox','@ngx-formly/material/select','exports', '@angular/core', '@ngx-formly/material/form-field', '@ngx-formly/material/input', '@ngx-formly/material/textarea', '@ngx-formly/material/radio', '@ngx-formly/material/checkbox', '@ngx-formly/material/multicheckbox', '@ngx-formly/material/select'], factory) :
    (factory(global.ng.core,global.ngxFormly.material.formField,global.ngxFormly.material.input,global.ngxFormly.material.textarea,global.ngxFormly.material.radio,global.ngxFormly.material.checkbox,global.ngxFormly.material.multicheckbox,global.ngxFormly.material.select,(global['ngx-formly'] = global['ngx-formly'] || {}, global['ngx-formly'].material = {}),global.ng.core,global['ngx-formly'].material['form-field'],global['ngx-formly'].material.input,global['ngx-formly'].material.textarea,global['ngx-formly'].material.radio,global['ngx-formly'].material.checkbox,global['ngx-formly'].material.multicheckbox,global['ngx-formly'].material.select));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,exports,core,formField,input,textarea,radio,checkbox,multicheckbox,select) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormlyMaterialModule = /** @class */ (function () {
        function FormlyMaterialModule() {
        }
FormlyMaterialModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: FormlyMaterialModule });
FormlyMaterialModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function FormlyMaterialModule_Factory(t) { return new (t || FormlyMaterialModule)(); }, imports: [[
            formField.FormlyMatFormFieldModule,
            input.FormlyMatInputModule,
            textarea.FormlyMatTextAreaModule,
            radio.FormlyMatRadioModule,
            checkbox.FormlyMatCheckboxModule,
            multicheckbox.FormlyMatMultiCheckboxModule,
            select.FormlyMatSelectModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FormlyMaterialModule, { imports: [ɵngcc1.FormlyMatFormFieldModule, ɵngcc2.FormlyMatInputModule, ɵngcc3.FormlyMatTextAreaModule, ɵngcc4.FormlyMatRadioModule, ɵngcc5.FormlyMatCheckboxModule, ɵngcc6.FormlyMatMultiCheckboxModule, ɵngcc7.FormlyMatSelectModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormlyMaterialModule, [{
        type: core.NgModule,
        args: [{
                imports: [
                    formField.FormlyMatFormFieldModule,
                    input.FormlyMatInputModule,
                    textarea.FormlyMatTextAreaModule,
                    radio.FormlyMatRadioModule,
                    checkbox.FormlyMatCheckboxModule,
                    multicheckbox.FormlyMatMultiCheckboxModule,
                    select.FormlyMatSelectModule,
                ]
            }]
    }], function () { return []; }, null); })();
        return FormlyMaterialModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.FieldType = formField.FieldType;
    exports.FormlyMaterialModule = FormlyMaterialModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
