// @ngx-formly/material/bundles/ngx-formly-material-slider.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/material/slider'),require('@angular/forms'),require('@ngx-formly/core'),require('@angular/common'),require('@ngx-formly/material/form-field'),exports, require('@angular/common'), require('@ngx-formly/core'), require('@angular/forms'), require('@angular/core'), require('@ngx-formly/material/form-field'), require('@angular/material/slider')) :
    typeof define === 'function' && define.amd ? define('@ngx-formly/material/slider', ['@angular/core','@angular/material/slider','@angular/forms','@ngx-formly/core','@angular/common','@ngx-formly/material/form-field','exports', '@angular/common', '@ngx-formly/core', '@angular/forms', '@angular/core', '@ngx-formly/material/form-field', '@angular/material/slider'], factory) :
    (factory(global.ng.core,global.ng.material.slider,global.ng.forms,global.ngxFormly.core,global.ng.common,global.ngxFormly.material.formField,(global['ngx-formly'] = global['ngx-formly'] || {}, global['ngx-formly'].material = global['ngx-formly'].material || {}, global['ngx-formly'].material.slider = {}),global.ng.common,global.core,global.ng.forms,global.ng.core,global['ngx-formly'].material['form-field'],global.ng.material.slider));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,exports,common,core,forms,core$1,formField,slider) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormlySliderTypeComponent = /** @class */ (function (_super) {
        __extends(FormlySliderTypeComponent, _super);
        function FormlySliderTypeComponent() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.defaultOptions = {
                templateOptions: {
                    hideFieldUnderline: true,
                    floatLabel: 'always',
                    thumbLabel: false,
                },
            };
            return _this;
        }
        /**
         * @param {?} event
         * @return {?}
         */
        FormlySliderTypeComponent.prototype.onContainerClick = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.slider.focus();
                _super.prototype.onContainerClick.call(this, event);
            };
        FormlySliderTypeComponent.propDecorators = {
            slider: [{ type: core$1.ViewChild, args: [slider.MatSlider,] }]
        };
FormlySliderTypeComponent.ɵfac = function FormlySliderTypeComponent_Factory(t) { return ɵFormlySliderTypeComponent_BaseFactory(t || FormlySliderTypeComponent); };
FormlySliderTypeComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FormlySliderTypeComponent, selectors: [["formly-field-mat-slider"]], viewQuery: function FormlySliderTypeComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(slider.MatSlider, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.slider = _t.first);
    } }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 1, vars: 11, consts: [[3, "id", "formControl", "formlyAttributes", "tabIndex", "color", "thumbLabel", "step", "max", "min"]], template: function FormlySliderTypeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "mat-slider", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("width", "100%");
        ɵngcc0.ɵɵproperty("id", ctx.id)("formControl", ctx.formControl)("formlyAttributes", ctx.field)("tabIndex", ctx.to.tabindex)("color", ctx.to.color)("thumbLabel", ctx.to.thumbLabel)("step", ctx.to.step)("max", ctx.to.max)("min", ctx.to.min);
    } }, directives: [ɵngcc1.MatSlider, ɵngcc2.NgControlStatus, ɵngcc2.FormControlDirective, ɵngcc3.FormlyAttributes], encapsulation: 2 });
var ɵFormlySliderTypeComponent_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(FormlySliderTypeComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormlySliderTypeComponent, [{
        type: core$1.Component,
        args: [{
                selector: 'formly-field-mat-slider',
                template: "\n    <mat-slider\n      [id]=\"id\"\n      [style.width]=\"'100%'\"\n      [formControl]=\"formControl\"\n      [formlyAttributes]=\"field\"\n      [tabIndex]=\"to.tabindex\"\n      [color]=\"to.color\"\n      [thumbLabel]=\"to.thumbLabel\"\n      [step]=\"to.step\"\n      [max]=\"to.max\"\n      [min]=\"to.min\">\n    </mat-slider>\n  "
            }]
    }], null, { slider: [{
            type: core$1.ViewChild,
            args: [slider.MatSlider]
        }] }); })();
        return FormlySliderTypeComponent;
    }(formField.FieldType));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FormlyMatSliderModule = /** @class */ (function () {
        function FormlyMatSliderModule() {
        }
FormlyMatSliderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: FormlyMatSliderModule });
FormlyMatSliderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function FormlyMatSliderModule_Factory(t) { return new (t || FormlyMatSliderModule)(); }, imports: [[
            common.CommonModule,
            forms.ReactiveFormsModule,
            slider.MatSliderModule,
            formField.FormlyMatFormFieldModule,
            core.FormlyModule.forChild({
                types: [{
                        name: 'slider',
                        component: FormlySliderTypeComponent,
                        wrappers: ['form-field']
                    }]
            }),
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FormlyMatSliderModule, { declarations: [FormlySliderTypeComponent], imports: [ɵngcc4.CommonModule, ɵngcc2.ReactiveFormsModule, ɵngcc1.MatSliderModule, ɵngcc5.FormlyMatFormFieldModule, ɵngcc3.FormlyModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormlyMatSliderModule, [{
        type: core$1.NgModule,
        args: [{
                declarations: [FormlySliderTypeComponent],
                imports: [
                    common.CommonModule,
                    forms.ReactiveFormsModule,
                    slider.MatSliderModule,
                    formField.FormlyMatFormFieldModule,
                    core.FormlyModule.forChild({
                        types: [{
                                name: 'slider',
                                component: FormlySliderTypeComponent,
                                wrappers: ['form-field']
                            }]
                    }),
                ]
            }]
    }], function () { return []; }, null); })();
        return FormlyMatSliderModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.FormlyMatSliderModule = FormlyMatSliderModule;
    exports.ɵa = FormlySliderTypeComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
