// @ngx-formly/material/esm5/form-field/form-field.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormlyModule } from '@ngx-formly/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatFormFieldModule } from '@angular/material/form-field';
import { FormlyWrapperFormField } from './form-field.wrapper';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@ngx-formly/core';
var FormlyMatFormFieldModule = /** @class */ (function () {
    function FormlyMatFormFieldModule() {
    }
FormlyMatFormFieldModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: FormlyMatFormFieldModule });
FormlyMatFormFieldModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function FormlyMatFormFieldModule_Factory(t) { return new (t || FormlyMatFormFieldModule)(); }, imports: [[
            CommonModule,
            ReactiveFormsModule,
            MatFormFieldModule,
            FormlyModule.forChild({
                wrappers: [{
                        name: 'form-field',
                        component: FormlyWrapperFormField
                    }]
            }),
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FormlyMatFormFieldModule, { declarations: function () { return [FormlyWrapperFormField]; }, imports: function () { return [CommonModule,
        ReactiveFormsModule,
        MatFormFieldModule, ɵngcc1.FormlyModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormlyMatFormFieldModule, [{
        type: NgModule,
        args: [{
                declarations: [FormlyWrapperFormField],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    MatFormFieldModule,
                    FormlyModule.forChild({
                        wrappers: [{
                                name: 'form-field',
                                component: FormlyWrapperFormField
                            }]
                    }),
                ]
            }]
    }], function () { return []; }, null); })();
    return FormlyMatFormFieldModule;
}());
export { FormlyMatFormFieldModule };
