// @ngx-formly/material/esm2015/slider/slider.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild } from '@angular/core';
import { FieldType } from '@ngx-formly/material/form-field';
import { MatSlider } from '@angular/material/slider';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/material/slider';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from '@ngx-formly/core';
export class FormlySliderTypeComponent extends FieldType {
    constructor() {
        super(...arguments);
        this.defaultOptions = {
            templateOptions: {
                hideFieldUnderline: true,
                floatLabel: 'always',
                thumbLabel: false,
            },
        };
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onContainerClick(event) {
        this.slider.focus();
        super.onContainerClick(event);
    }
}
FormlySliderTypeComponent.ɵfac = function FormlySliderTypeComponent_Factory(t) { return ɵFormlySliderTypeComponent_BaseFactory(t || FormlySliderTypeComponent); };
FormlySliderTypeComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FormlySliderTypeComponent, selectors: [["formly-field-mat-slider"]], viewQuery: function FormlySliderTypeComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(MatSlider, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.slider = _t.first);
    } }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 1, vars: 11, consts: [[3, "id", "formControl", "formlyAttributes", "tabIndex", "color", "thumbLabel", "step", "max", "min"]], template: function FormlySliderTypeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "mat-slider", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵstyleProp("width", "100%");
        ɵngcc0.ɵɵproperty("id", ctx.id)("formControl", ctx.formControl)("formlyAttributes", ctx.field)("tabIndex", ctx.to.tabindex)("color", ctx.to.color)("thumbLabel", ctx.to.thumbLabel)("step", ctx.to.step)("max", ctx.to.max)("min", ctx.to.min);
    } }, directives: [ɵngcc1.MatSlider, ɵngcc2.NgControlStatus, ɵngcc2.FormControlDirective, ɵngcc3.FormlyAttributes], encapsulation: 2 });
FormlySliderTypeComponent.propDecorators = {
    slider: [{ type: ViewChild, args: [MatSlider,] }]
};
const ɵFormlySliderTypeComponent_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(FormlySliderTypeComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormlySliderTypeComponent, [{
        type: Component,
        args: [{
                selector: 'formly-field-mat-slider',
                template: `
    <mat-slider
      [id]="id"
      [style.width]="'100%'"
      [formControl]="formControl"
      [formlyAttributes]="field"
      [tabIndex]="to.tabindex"
      [color]="to.color"
      [thumbLabel]="to.thumbLabel"
      [step]="to.step"
      [max]="to.max"
      [min]="to.min">
    </mat-slider>
  `
            }]
    }], null, { slider: [{
            type: ViewChild,
            args: [MatSlider]
        }] }); })();
if (false) {
    /** @type {?} */
    FormlySliderTypeComponent.prototype.slider;
    /** @type {?} */
    FormlySliderTypeComponent.prototype.defaultOptions;
}
