// @ngx-formly/material/esm5/slider/slider.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormlyModule } from '@ngx-formly/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatSliderModule } from '@angular/material/slider';
import { FormlyMatFormFieldModule } from '@ngx-formly/material/form-field';
import { FormlySliderTypeComponent } from './slider.type';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@ngx-formly/core';
var FormlyMatSliderModule = /** @class */ (function () {
    function FormlyMatSliderModule() {
    }
FormlyMatSliderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: FormlyMatSliderModule });
FormlyMatSliderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function FormlyMatSliderModule_Factory(t) { return new (t || FormlyMatSliderModule)(); }, imports: [[
            CommonModule,
            ReactiveFormsModule,
            MatSliderModule,
            FormlyMatFormFieldModule,
            FormlyModule.forChild({
                types: [{
                        name: 'slider',
                        component: FormlySliderTypeComponent,
                        wrappers: ['form-field']
                    }]
            }),
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FormlyMatSliderModule, { declarations: function () { return [FormlySliderTypeComponent]; }, imports: function () { return [CommonModule,
        ReactiveFormsModule,
        MatSliderModule,
        FormlyMatFormFieldModule, ɵngcc1.FormlyModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormlyMatSliderModule, [{
        type: NgModule,
        args: [{
                declarations: [FormlySliderTypeComponent],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    MatSliderModule,
                    FormlyMatFormFieldModule,
                    FormlyModule.forChild({
                        types: [{
                                name: 'slider',
                                component: FormlySliderTypeComponent,
                                wrappers: ['form-field']
                            }]
                    }),
                ]
            }]
    }], function () { return []; }, null); })();
    return FormlyMatSliderModule;
}());
export { FormlyMatSliderModule };
